rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java

```diff
@@ -1570,15 +1571,48 @@ private void adjustBalancesIfChangeOutputWasDust(BtcTransaction btcTx, Coin sent
      * The hash for the signature must be calculated with Transaction.SigHash.ALL and anyoneCanPay=false. The signature must be canonical.
      * If enough signatures were added, ask federators to broadcast the btc release tx.
      *
-     * @param federatorPublicKey   Federator who is signing
+     * @param federatorBtcPublicKey   Federator who is signing
      * @param signatures           1 signature per btc tx input
-     * @param rskTxHash            The id of the rsk tx
+     * @param rskTxHashSerialized            The id of the rsk tx
      */
-    public void addSignature(BtcECKey federatorPublicKey, List<byte[]> signatures, byte[] rskTxHash) throws Exception {
+    public void addSignature(BtcECKey federatorBtcPublicKey, List<byte[]> signatures, byte[] rskTxHashSerialized) throws Exception {
+        if (signatures == null || signatures.isEmpty()) {
+            return;
+        }
+
         Context.propagate(btcContext);
 
+        Keccak256 rskTxHash = new Keccak256(rskTxHashSerialized);
+        int signaturesSize = signatures.size();
+
+        BtcTransaction btcTx = provider.getPegoutsWaitingForSignatures().get(rskTxHash);
+        if (btcTx == null) {
+            logger.warn("No tx waiting for signature for hash {}. Probably fully signed already.", rskTxHash);
+            return;
+        }
+        if (notEnoughSignatures(btcTx.getInputs().size(), signaturesSize)) {
+            return;
+        }
+
+        addReleaseSignatures(federatorBtcPublicKey, signatures, rskTxHashSerialized, btcTx);
+    }
+
+    private boolean notEnoughSignatures(int inputsSize, int signaturesSize) {
+        if (inputsSize != signaturesSize) {
+            logger.warn("Expected {} signatures but received {}.", inputsSize, signaturesSize);
+            return true;
+        }
+        return false;
+    }
+
+    private void addReleaseSignatures(
+        BtcECKey federatorPublicKey,
+        List<byte[]> signatures,
+        byte[] rskTxHashSerialized,
```

Comment:
Maybe it would be a good idea to replace this by the deserialized `Keccak256` hash

Reply:
it uses both the serialized and the deserialized one, do you see any advantages in passing the deserialized one?

Reply:
Maybe with the deserialized one you can do something like `getBytes()` or something. I am saying this to avoid doing it twice

Reply:
in that case we are doing `rskTxHashSerialized -> rskTxHash -> rskTxHash.getBytes()`
instead of just `rskTxHashSerialized`